beacon_node/operation_pool/src/reward_cache.rs

```diff
@@ -0,0 +1,125 @@
+    /// `BitVec` of validator indices which don't have default participation flags for the current epoch.
+    current_epoch_participation: BitVec,
+}
+
+impl RewardCache {
+    pub fn has_attested_in_epoch(
+        &self,
+        validator_index: u64,
+        epoch: Epoch,
+    ) -> Result<bool, OpPoolError> {
+        if let Some(init) = &self.initialization {
+            if init.current_epoch == epoch {
+                Ok(*self
+                    .current_epoch_participation
+                    .get(validator_index as usize)
+                    .ok_or(OpPoolError::RewardCacheOutOfBounds)?)
+            } else if init.current_epoch == epoch + 1 {
+                Ok(*self
+                    .previous_epoch_participation
+                    .get(validator_index as usize)
+                    .ok_or(OpPoolError::RewardCacheOutOfBounds)?)
+            } else {
+                Err(OpPoolError::RewardCacheWrongEpoch)
+            }
+        } else {
+            Err(OpPoolError::RewardCacheWrongEpoch)
+        }
+    }
+
+    // Determine the "marker" block root to store in `self.init` for a given `slot`.
+    //
+    // For simplicity at genesis we return the zero hash, which will cause one unnecessary
+    // re-calculation.
+    fn marker_block_root<E: EthSpec>(
+        state: &BeaconState<E>,
+        slot: Slot,
+    ) -> Result<Hash256, OpPoolError> {
+        if slot == 0 {
+            Ok(Hash256::zero())
+        } else {
+            Ok(*state
+                .get_block_root(slot)
+                .map_err(OpPoolError::RewardCacheGetBlockRoot)?)
+        }
+    }
+
+    /// Update the cache.
+    pub fn update<E: EthSpec>(&mut self, state: &BeaconState<E>) -> Result<(), OpPoolError> {
+        if state.previous_epoch_participation().is_err() {
+            return Ok(());
+        }
+
+        let current_epoch = state.current_epoch();
+        let latest_block_root = Self::marker_block_root(state, state.slot() - 1)?;
+
+        let new_init = Initialization {
+            current_epoch,
+            latest_block_root,
+        };
+
+        // The participation flags change every block, and will almost always need updating when
+        // this function is called at a new slot.
+        if self
+            .initialization
+            .as_ref()
+            .map_or(true, |init| *init != new_init)
+        {
+            self.update_previous_epoch_participation(state)
+                .map_err(OpPoolError::RewardCacheUpdatePrevEpoch)?;
+            self.update_current_epoch_participation(state)
+                .map_err(OpPoolError::RewardCacheUpdateCurrEpoch)?;
+
+            self.initialization = Some(new_init);
+        }
+
+        Ok(())
+    }
+
+    fn update_previous_epoch_participation<E: EthSpec>(
+        &mut self,
+        state: &BeaconState<E>,
+    ) -> Result<(), BeaconStateError> {
+        let default_participation = ParticipationFlags::default();
+        self.previous_epoch_participation = state
```

Comment:
The idea for tree-states is to pay the iteration cost once while converting the participation flags to a form with `O(1)` indexing (rather than `O(log n)`). So cloning the `VariableList` of flags as-is isn't viable, and if we're iterating the whole thing we _may as well_ smush it into a tiny bitvec for later use.

Reply:
Makes sense to me :)